### ISSUE TITLE
How to properly create a Helm chart for operator?

### ISSUE THREAD

**Issue by ninja-** [operator-framework/operator-sdk#4930]
## Type of question

General operator-related help

#### What did you do?

I am looking to upgrade Operator SDK after about a ~ 1.5 year of working on previous version, and I am a bit confused.
How can I package my operator as a normal helm chart?
I see no mentions of that anywhere in the docs.
I tried running `make bundle`, but resulting manifests are removing **Deployment** objects every time, no matter if I include them manually into kustomize config.
Instead I get a 
```
apiVersion: operators.coreos.com/v1alpha1
kind: ClusterServiceVersion
```

object which I am not interested in, as I understand that's some core "operator for installing operator", but I would like to stick to plain old deployment objects, which `make bundle` is cutting out.

#### Environment

**Operator type:**

/language go

**Kubernetes cluster type:**

`$ operator-sdk version`

`$ go version` (if language is Go)

`$ kubectl version`

**Comment by varshaprasad96** [comment]
@ninja- that's the right step. After creating an operator project out of helm chart, you can run `make bundle` to package your operator. The deployment should be have been specified in CSV, do you mean its not present and being removed? [CSV](https://olm.operatorframework.io/docs/concepts/crds/clusterserviceversion/#writing-your-operator-manifests) is used by OLM to deploy an instance of an operator.

**Comment by varshaprasad96** [comment]
@ninja- Please ignore my previous comment! Did you mean - you would like to package a `go` operator in a helm chart to manage its lifecycle? That should be possible, and does not need any additional steps from SDK's end. Just to clarify here, [`make bundle`](https://sdk.operatorframework.io/docs/olm-integration/quickstart-bundle/) is used to package an operator for Operator Lifecycle Manager (OLM).

I am just curious to know more about the use case of packaging the operator project into helm chart here. If you are looking for managing the operator and its resources on a cluster, [Operator Lifecycle Manager](https://olm.operatorframework.io) is a good candidate. You can directly package and deploy the go operator and leave the management to olm. 

If you would like to try using OLM, here are some links which can guide you through packaging the operator bundle and deploying it using OLM:
1. Creating operator bundle: https://sdk.operatorframework.io/docs/olm-integration/quickstart-bundle/
2. Installing OLM on cluster: https://sdk.operatorframework.io/docs/olm-integration/quickstart-bundle/#enabling-olm
3. Getting started with OLM: https://olm.operatorframework.io/docs/getting-started/
4. More on OLM and its capabilities: https://olm.operatorframework.io/docs/#features-provided-by-olm

**Comment by ninja-** [comment]
@varshaprasad96 I am looking to simply create a helm chart that's able to install my operator without OLM on the cluster, and so no ClusterServiceVersion object. Just plain old Deployment, Role, ServiceAccount & CRDs.
I was able to hack this around and create these manifests, but it was a long manual process that could have been integrated some way into operator-sdk and kustomize templates.
Currently no matter how hard I try to include Deployment objects into kustomize config, they get cut out by `make bundle`

**Comment by varshaprasad96** [comment]
@ninja- Can you please elaborate more on why you would like to use helm chart?
`make bundle` is specific for packaging an operator for OLM. That cannot be used for creating helm chart for the operator.

**Comment by ninja-** [comment]
Yes it is specific, but what I want to do is have a kustomize base file that includes needed objects for a pure helm chart (which doesn't differ much from the default config used today that makes the _bundle_) and then split the objects into several files, which kustomize can't do, and that's handled by operator-sdk cli

**Comment by estroz** [comment]
The `operator-sdk` CLI will only split up those files in the context of an OLM bundle, and won't really do what you want. What I suggest is using something like `kustomize build config/default | [yq eval](https://github.com/mikefarah/yq)` to process these into multiple docs.

**Comment by ninja-** [comment]
that's more or less what I did at this point. but it is still very hacky and would be nice it was integrated with operator-sdk instead.

**Comment by estroz** [comment]
Typically either helm or kustomize are used to "fill in" some values in k8s manifests, but not both (except using kustomize as a helm [post-renderer](https://github.com/thomastaylor312/advanced-helm-demos/tree/master/post-render)). A helm template is not a valid k8s manifest, and will error if you try to `kustomize build` them in some cases. These concepts do not sound foreign to you but I wanted to state them just in case.

Given the above, I don't really see a reason why `operator-sdk` would be a useful intermediary, since you are effectively asking that `operator-sdk` act as a script to split `kustomize build` output and tons of tools already do this. It would be better to re-configure your operator as helm-first rather than use kustomize at all. Otherwise, what's the point of using helm in the first place instead of `kustomize build config/default | kubectl apply -f -`? You would be missing out on a bunch of benefits that helm charts get. Perhaps I'm missing something about your use case.

**Comment by ninja-** [comment]
I don't understand the `kubectl apply` vs `helm` argument in this case.
The manifests rendered by `kustomize` are still generic enough such that they can be thrown into a helm chart without much problems, with only minor changes.
If I do any adjustments to fit helm, these are post `kustomize` and not before, if that's what you are referring to.

Right now I am using the operator-sdk automation that updates CRDs and RBAC permissions off the source code changes, so right now that is the pain point to keep in sync when maintaining a non-CSV helm chart.

**Comment by estroz** [comment]
Ok thanks for clarifying.

>I don't understand the kubectl apply vs helm argument in this case.

I'm saying that `kustomize` and `helm` are really two different templating formats, the latter of which also doubles as a packaging format. So you're kinda losing the benefit of the templating side of helm by trying to use kustomize as the templater. You still certainly can, but I don't think doing so is recommended, nor should operator-sdk should be in the business of bridging this gap. I'm still a little confused as to why you think splitting files output by `kustomize build` would be better done by `operator-sdk` than by a simple script. The latter is a lot more flexible.

>Right now I am using the operator-sdk automation that updates CRDs and RBAC permissions off the source code changes, so right now that is the pain point to keep in sync when maintaining a non-CSV helm chart.

All you need to do to update these files from source code markers is run `make manifests`. You could basically create a make target along the following lines:
```make
build-helm: generate manifests
	mkdir -p helm
	kustomize build config/default | awk '{f="helm/" NR "_manifest.yaml"; print $0 > f}' RS='\n---\n'
```
The script can probably be better at naming the files but this is all possible within the scripting language of your choice.

**Comment by ninja-** [comment]
> All you need to do to update these files from source code markers is run make manifests. You could basically create a make target along the following lines:

yes, that's more or less what I have done already, but it still feels more hacky than it should be

**Comment by mudler** [comment]
I've stumbled into this myself and I was surprised to see it's not bundled somehow in the generated Makefile.

I'm not a fun of scripting solutions, so I've quickly assembled a small utility, if anyone else have the same issue, there you go: https://github.com/spectrocloud/kubesplit

**Comment by abdennour** [comment]
Use Helmify https://github.com/arttor/helmify

```Makefile
HELMIFY = $(shell pwd)/bin/helmify
helmify:
	$(call go-get-tool,$(HELMIFY),github.com/arttor/helmify/cmd/helmify@v0.3.7)

helm: manifests kustomize helmify
	$(KUSTOMIZE) build config/default | $(HELMIFY)
```

**Comment by Praveenk8051** [comment]
I stumbled on this issue too.
Maybe one can do like this..

`apiVersion: batch/v1
kind: Job
metadata:
  name: {{ include "my-chart.fullname" . }}-make-install
  labels:
    {{- include "my-chart.labels" . | nindent 4 }}
spec:
  template:
    spec:
      containers:
      - name: make-install
        image: alpine:3.14
        command: ["sh", "-c", "apk add --no-cache make && cd /src && make install"]
        volumeMounts:
        - name: source-code
          mountPath: /src
      restartPolicy: Never
      volumes:
      - name: source-code
        gitRepo:
          repository: "https://github.com/your/repo.git"
          revision: "master"
`

Would love to know your thoughts!
@ninja-

**Comment by Praveenk8051** [comment]
Update - [Helmify](https://github.com/arttor/helmify) Works too good
